### DISCUSSION TITLE
Disable Trimming For Entire Project/Assembly

### DISCUSSION THREAD

**Discussion by SerratedSharp** [dotnet/runtime#95891] [Q&A]
Since losing default constructors due to trimming causes alot of problems for deserializers and my workaround has been sprinkling `[DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(AnotherModel))]` all over the place.  

So I'm trying to isolate all my serializable types into a class library project and prevent trimming on that assembly as a whole.  I thought changing TrimMode to partial on the console app then setting IsTrimmable to false on the class library would prevent it from being trimmed.

On the console app (that is compiled to a WebAssembly with Uno Boostrap) I added this to the *.csproj:
```
<PropertyGroup>
		<TrimMode>partial</TrimMode>
</PropertyGroup>
```

I added a AssemblyInfo.cs to the class library with this:
`[assembly: AssemblyMetadata("IsTrimmable", "False")]`

Being thorough I also added this to the class library's csproj:
```
<PropertyGroup>
	  <IsTrimmable>false</IsTrimmable>
  </PropertyGroup>
```

I assume I'm misunderstanding the documentation.  Is there a way to disable trimming as a whole for the class library while leaving it enabled in the console app?

**Comment by MichalStrehovsky** [comment]
The .NET team's recommendation is to switch to a serializer that is compatible with trimming. As you found out, the experience is going to be very unpleasant and the publish-time warnings generated for the reflection-based serializer will cause a lot of noise. You'll likely miss things due to the warning noise. Quoting from [the docs](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/fixing-warnings): "If there are any trim warnings, the app should be thoroughly tested after trimming to ensure that there are no behavior changes." 

That said, if you set `<TrimMode>` to `partial`, any assembly that is not already marked as IsTrimmable=true will not be trimmed. You don't need to do `<IsTrimmable>false</IsTrimmable>` because it's the implicit behavior; setting IsTrimmable to false doesn't do anything.

If you want to keep an assembly in it's entirety and trim everything else, leave `<TrimMode>` at it's default (`full`) in the console project, and the add `<TrimmerRootAssembly Include="NameOfAssemblyYouWantToKeep" />` (also in the console project). That will trim everything, but keep things in that one specific assembly.

**Reply by SerratedSharp** [reply]
Thanks, I do plan to migrate serializers eventually, but having done the Newtonsoft->System.Text migration in other projects, for what I'm doing right now I'm willing to accept a larger assembly size for the particular project holding my POCOs, in order to avoid going down that rabbit hole.

In my WASM console app I updated my project reference to the "ClientModels" Class Library like so:
```XML
	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net8.0</TargetFramework>
		<TrimMode>full</TrimMode>
	</PropertyGroup>
	<ItemGroup>		
		<ProjectReference Include="..\ClientModels\ClientModels.csproj" />		
		<TrimmerRootAssembly Include="ClientModels" />
	</ItemGroup>
```

Am I doing this right for a project reference or am I misunderstanding what I expect it would do?  I'm thinking all classes inside ClientModels would have their members including constructors preserved.

 I created a new class library just to make sure namespace, DLL name, and assembly name property were all aligned. **I still get NoDefaultConstructor errors on the root type/POCO encountered by the Deserializer.**  Maybe I'm misunderstanding what this would do, or I'm confused about which IL Linker is leveraged by Uno.Bootstrap when I set WasmShellILLinkerEnabled.  I know at one time the IL Linker for bootstrap was a separate bit of tooling aside from .NET, but I thought that since at least .NET 7 they were using the .NET runtime linker.  So maybe I'm in the wrong place.

As a sanity check I can get it working successfully by calling out the POCOs with `[DynamicDependency...]` or wrapping my POCOs in a class so that they are all nested classes of a `TrimmerDMZ` class, and then doing `[DynamicDependency(DynamicallyAccessedMemberTypes.PublicNestedTypes, typeof(TrimmerDMZ))]`

**Reply by SerratedSharp** [reply]
**My apologies**, I guess I'm in the wrong place. It seems Uno.Bootstrap 8 is still using this linker: https://github.com/dotnet/linker/

I added this to my WASM client:
```
<ItemGroup>
    <LinkerDescriptor Include="LinkerConfig.xml" />
</ItemGroup>
```

With the following in the LinkerConfig.xml and it suppressed trimming for my POCOs.
```
<linker>
	<assembly fullname="ClientModels" preserve="all" />
</linker>
```

**Comment by alexandrehtrb** [comment]
Writing here because I am a bit confused about trimming.

Consider the following dependency graph, A is the root assembly; B and C are direct dependencies; X,Y and Z are transitive dependencies:

```mermaid
graph LR
  A-->B-->X
  A-->C-->Y
  C-->Z
  
  style A fill:#ADF,stroke:#333,stroke-width:4px,color:#000
  style B fill:#7F9,stroke:#333,stroke-width:4px,color:#000
  style C fill:#7F9,stroke:#333,stroke-width:4px,color:#000
  style X fill:#f9f,stroke:#333,stroke-width:4px,color:#000
  style Y fill:#f9f,stroke:#333,stroke-width:4px,color:#000
  style Z fill:#7F9,stroke:#333,stroke-width:4px,color:#000
```

I want to remove X and Y, because I am certain that they will not be needed. A, B, C and Z must remain.

How can I trim assembly A for that to happen?

I tried the following in `A.csproj`, but a lot of warnings popped up, unrelated to X and Y.

```xml
<PropertyGroup>
  <AssemblyName>A</AssemblyName>
  <PublishTrimmed>True</PublishTrimmed>
  <TrimMode>partial</TrimMode> 
</PropertyGroup>
<ItemGroup>
  <!-- these are the assemblies that will be analysed to be trimmed -->
  <TrimmableAssembly Include="X" />
  <TrimmableAssembly Include="Y" />
</ItemGroup>
<ItemGroup>
  <PackageReference Include="B" Version="1.0.0" />
  <PackageReference Include="C" Version="1.0.0" />
</ItemGroup>
```

**Reply by MichalStrehovsky** [reply]
> How can I trim assembly A for that to happen?

You just did that. TrimMode partial means that only assemblies marked as trimmable will be trimmed. Provided that B, C and A are not declared as such (can't tell this from your csproj - they could be marked that way in the assembly metadata), they will be kept.

The warning says that there is code that could not be statically analyzed. If you're confident that what you just did addresses all the potential issues, you can just disable the warning with `<SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>` in the project.

Note that if you do that, you assume the full responsibility that the trimmed app will behave as intended (if it gets broken by trimming, it's not a bug in trimming, it's a bug in your project).

Our recommendation is to use `TrimMode=full` and _address_ (not _suppress_) the warnings. It may involve having to work with third parties if the warnings are in third party code.

**Reply by SerratedSharp** [reply]
I would have thought based on the documentation's "can opt-in individual assemblies to trimming by using a `<TrimmableAssembly>`" that this would perform the analysis based trimming from the perspective of the root.  If instead you're saying it removes those assemblies explicitly and entirely without regard to analysis then feel like the documentation needs stronger wording.

**Reply by alexandrehtrb** [reply]
@SerratedSharp , it isn't meant to remove the assembly entirely; what happens is that the trimming process will scan it and notice it isn't used, and by that, it will mostly be eliminated due to trimming unused code.

(that's what I understood)
